File: DEVELOPMENT/Python/ground detection/C translation/solidity_detection.c
/*
 * @file modules/computer_vision/solidity_detection.c
 * @brief Blob shape analysis for ground vs obstacle classification
 *
 * Translated from solidity_detection.py
 * Uses perimeter ratio and fractal dimension to determine if a blob
 * is smooth (ground) or irregular (plant/obstacle).
 */

#include "solidity_detection.h"
#include <math.h>
#include <string.h>
#include <stdio.h>

// -----------------------------------------------------------------------
// get_blob_edge
// Marks pixels where a transition between 0 and non-zero occurs.
// Equivalent to get_blob_edge() in solidity_detection.py
//
// edge_out must be same size as binary_img (w * h bytes)
// -----------------------------------------------------------------------
void get_blob_edge(uint8_t *binary_img, uint8_t *edge_out,
                   uint16_t w, uint16_t h)
{
  memset(edge_out, 0, w * h);

  // Horizontal transitions
  for (uint16_t row = 0; row < h; row++) {
    for (uint16_t col = 0; col < w - 1; col++) {
      uint8_t curr  = binary_img[row * w + col]     > 0 ? 1 : 0;
      uint8_t right = binary_img[row * w + col + 1] > 0 ? 1 : 0;
      if (curr != right) {
        edge_out[row * w + col] = 1;
      }
    }
  }

  // Vertical transitions
  for (uint16_t row = 0; row < h - 1; row++) {
    for (uint16_t col = 0; col < w; col++) {
      uint8_t curr = binary_img[row * w + col]       > 0 ? 1 : 0;
      uint8_t down = binary_img[(row + 1) * w + col] > 0 ? 1 : 0;
      if (curr != down) {
        edge_out[row * w + col] = 1;
      }
    }
  }
}

// -----------------------------------------------------------------------
// blob_perimeter
// Counts transitions between 0 and non-zero pixels horizontally
// and vertically.
// Equivalent to blob_perimeter() in solidity_detection.py
//
// NOTE: mask must be padded by 1 pixel on all sides (w and h include padding)
// -----------------------------------------------------------------------
uint32_t blob_perimeter(uint8_t *padded_mask, uint16_t w, uint16_t h)
{
  uint32_t perimeter = 0;

  // Horizontal transitions
  for (uint16_t row = 0; row < h; row++) {
    for (uint16_t col = 0; col < w - 1; col++) {
      uint8_t curr  = padded_mask[row * w + col]     > 0 ? 1 : 0;
      uint8_t right = padded_mask[row * w + col + 1] > 0 ? 1 : 0;
      if (curr != right) perimeter++;
    }
  }

  // Vertical transitions
  for (uint16_t row = 0; row < h - 1; row++) {
    for (uint16_t col = 0; col < w; col++) {
      uint8_t curr = padded_mask[row * w + col]       > 0 ? 1 : 0;
      uint8_t down = padded_mask[(row + 1) * w + col] > 0 ? 1 : 0;
      if (curr != down) perimeter++;
    }
  }

  return perimeter;
}

// -----------------------------------------------------------------------
// compute_average_blob_height
// Equivalent to compute_average_blob_height() in solidity_detection.py
// -----------------------------------------------------------------------
float compute_average_blob_height(float blob_area, uint16_t bounding_box_width)
{
  if (bounding_box_width == 0) return 0.0f;
  return blob_area / (float)bounding_box_width;
}

// -----------------------------------------------------------------------
// compute_fractal_dimension
// Box-counting fractal dimension of the blob edge.
// Equivalent to compute_fractal_dimension() in solidity_detection.py
// -----------------------------------------------------------------------
float compute_fractal_dimension(uint8_t *binary_img, uint16_t w, uint16_t h)
{
  // Get edge pixels
  uint8_t edge[IMAGE_WIDTH * IMAGE_HEIGHT];
  get_blob_edge(binary_img, edge, w, h);

  // Box sizes as powers of 2 up to min(w, h)
  uint16_t min_dim = (w < h) ? w : h;

  // Max 8 scales (2^1 to 2^8 = 256)
  float eps_axis[8];
  float box_axis[8];
  uint8_t n_scales = 0;

  for (uint8_t i = 1; (1 << i) <= min_dim; i++) {
    uint16_t s = 1 << i; // box size = 2^i

    uint16_t trimmed_h = (h / s) * s;
    uint16_t trimmed_w = (w / s) * s;

    uint32_t count = 0;

    // Count non-empty s x s boxes
    for (uint16_t br = 0; br < trimmed_h; br += s) {
      for (uint16_t bc = 0; bc < trimmed_w; bc += s) {
        bool any = false;
        for (uint16_t r = br; r < br + s && !any; r++) {
          for (uint16_t c = bc; c < bc + s && !any; c++) {
            if (edge[r * w + c] > 0) any = true;
          }
        }
        if (any) count++;
      }
    }

    if (count > 0) {
      eps_axis[n_scales] = logf((float)s);
      box_axis[n_scales] = logf((float)count);
      n_scales++;
    }
  }

  if (n_scales < 2) return 0.0f;

  // Linear regression (polyfit degree 1) on log-log data
  float sum_x = 0, sum_y = 0, sum_xy = 0, sum_x2 = 0;
  for (uint8_t i = 0; i < n_scales; i++) {
    sum_x  += eps_axis[i];
    sum_y  += box_axis[i];
    sum_xy += eps_axis[i] * box_axis[i];
    sum_x2 += eps_axis[i] * eps_axis[i];
  }

  float denom = (float)n_scales * sum_x2 - sum_x * sum_x;
  if (fabsf(denom) < 1e-10f) return 0.0f;

  float slope = ((float)n_scales * sum_xy - sum_x * sum_y) / denom;

  // Slope is negative, negate to get positive dimension
  return -slope;
}

// -----------------------------------------------------------------------
// is_smooth_blob
// Returns true if blob is smooth (ground), false if irregular (obstacle).
// Equivalent to is_smooth_blob() in solidity_detection.py
// -----------------------------------------------------------------------
bool is_smooth_blob(uint8_t *single_blob_mask, uint16_t w, uint16_t h,
                    float area, uint16_t bounding_box_width)
{
  // Pad mask by 1 pixel on all sides
  uint16_t pw = w + 2;
  uint16_t ph = h + 2;
  uint8_t padded[pw * ph];
  memset(padded, 0, pw * ph);

  for (uint16_t row = 0; row < h; row++) {
    for (uint16_t col = 0; col < w; col++) {
      padded[(row + 1) * pw + (col + 1)] = single_blob_mask[row * w + col];
    }
  }

  // Compute perimeter of padded blob
  uint32_t perimeter = blob_perimeter(padded, pw, ph);

  // Compute average height and equivalent rectangle perimeter
  float avg_height = compute_average_blob_height(area, bounding_box_width);
  float equiv_rect_perimeter = 2.0f * (avg_height + (float)bounding_box_width);

  // Compute fractal dimension
  float fractal_dim = compute_fractal_dimension(single_blob_mask, w, h);

  // Check smoothness conditions
  // If either condition is met -> blob is smooth -> return true
  if (equiv_rect_perimeter > 0.0f) {
    float perimeter_ratio = (float)perimeter / equiv_rect_perimeter;
    if (perimeter_ratio < THRESHOLD_PERIMETER_RATIO
        || fractal_dim < THRESHOLD_FRACTAL_DIM) {
      return true;
    }
  }

  return false;
}

File: DEVELOPMENT/Python/optical_flow/solidity_detection.py
import numpy as np
import cv2

def is_smooth_blob_old(single_blob_mask: np.ndarray,
                   area: float) -> bool:
    """
    Returns True if the blob in the mask is smooth (likely ground),
    False if it is spiky/irregular (likely plant).
    Uses only basic array operations — easy to port to C.
    
    Args:
        mask: binary image (uint8, 0 or 255) containing *ONE* blob
        area: blob area
    """
    solidity_thresh     = 0.8
    circularity_thresh  = 0.0

    # get a mask to identify the current blob (using the label)
    pixels = (single_blob_mask != 0)
    
    # pad image by 1 pixel
    padded = np.pad(pixels, 1, mode='constant', constant_values=0)

    perimeter = blob_perimeter(padded_binary_img=padded)

    # ys, xs = np.where(pixels)
    # points = np.stack([xs, ys], axis=1).astype(np.float32)
    # hull_area = convex_hull_area_fast(points)
    
    # # ---- SOLIDITY ----
    # if area > 0:
    #     solidity = area / hull_area
    # else:
    #     solidity = 0

    # --- CIRCULARITY ---
    if perimeter > 0:
        circularity = (4 * np.pi * area) / (perimeter ** 2)
    else: 
        circularity = 0

    # check_solidity = solidity > solidity_thresh
    check_circularity = circularity > circularity_thresh
    return check_circularity # and check_solidity


def is_smooth_blob(single_blob_mask: np.ndarray,
                                        area: float,
                                        bounding_box_width: int) -> bool:
    """
    Returns True if the blob in the mask is smooth (likely ground),
    False if it is spiky/irregular (likely plant).
    Uses only basic array operations — easy to port to C.
    
    Args:
        mask: binary image (uint8, 0 or 255) containing *ONE* blob
        area: blob area
        bb_width: bounding box WIDTH (from roof to ground)
        bb_height: bounding box HEIGHT (from left to right)
    
    NOTE: 
        bb width and height are assumed in an absolute sense. 
        If the image is rotated, like the ones produced by the Bebop,
        the width and height will be swapped.
    """

    # get a mask to identify the current blob (using the label)
    pixels = (single_blob_mask != 0)
    
    # pad image by 1 pixel
    padded = np.pad(pixels, 1, mode='constant', constant_values=0)

    perimeter = blob_perimeter(padded_binary_img=padded)

    # compute AVERAGE HEIGHT (per-column) of blob + equivalent rectangle perimeter
    average_height = compute_average_blob_height(area, bounding_box_width)
    equivalent_rectangle_perimeter = 2 * (average_height + bounding_box_width)

    # compute fractal dimension of blob
    fractal_dim = compute_fractal_dimension(padded)

    threshold_perimeter     = 2
    threshold_fractal_dim   = 1.3
    if perimeter / equivalent_rectangle_perimeter < threshold_perimeter \
                                  or fractal_dim < threshold_fractal_dim:
        return True

    return False


def compute_average_blob_height(blob_area: float,
                                bounding_box_width: int):
    """
    Computes the average blob height, considering an image rotated with the groun on the left (-90° rotation).
    Therefore, the blob height corresponds to the column-wise span of the image.
    Moreover, this assumes the image contains ONE single blob.
    """

    # reduce the image size to fit the blob perfectly.
    avg_height = blob_area // bounding_box_width

    return avg_height


def blob_perimeter(padded_binary_img: np.ndarray):#
    """
    Compute perimeter of a binary blob of pixels using vectorized NumPy operations.
    This assumes ONE blob per image.
    """
               
    perimeter = np.sum(padded_binary_img[:,1:] != padded_binary_img[:,:-1]) + \
                np.sum(padded_binary_img[1:,:] != padded_binary_img[:-1,:])

    return perimeter


def convex_hull_area_fast(points: np.ndarray) -> float:
    """
    Convex hull area via Monotone Chain + Shoelace.
    Args:
        points: (N, 2) array of 2D points
    Returns:
        convex hull area
    """
    pts = points[np.lexsort((points[:, 1], points[:, 0]))]
    pts = np.unique(pts, axis=0)
    n = len(pts)
    if n < 3:
        return 0.0

    def cross(o, a, b):
        return (a[0]-o[0]) * (b[1]-o[1]) - (a[1]-o[1]) * (b[0]-o[0])

    lower = []
    for p in pts:
        while len(lower) >= 2 and cross(lower[-2], lower[-1], p) <= 0:
            lower.pop()
        lower.append(p)

    upper = []
    for p in reversed(pts):
        while len(upper) >= 2 and cross(upper[-2], upper[-1], p) <= 0:
            upper.pop()
        upper.append(p)

    hull = np.array(lower[:-1] + upper[:-1])
    x, y = hull[:, 0], hull[:, 1]
    return 0.5 * abs(np.dot(x, np.roll(y, -1)) - np.dot(y, np.roll(x, -1)))


def convex_hull_area(points: np.ndarray) -> float:
    """
    Computes convex hull area using Graham scan + shoelace formula.
    Pure numpy/python -> portable to C.
    """
    # Find pivot: lowest y, then leftmost x
    idx = np.lexsort((points[:, 0], points[:, 1]))
    points = points[idx]
    pivot = points[0]

    # Sort by polar angle from pivot
    deltas = points[1:] - pivot
    angles = np.arctan2(deltas[:, 1], deltas[:, 0])
    dists  = deltas[:, 0]**2 + deltas[:, 1]**2
    order  = np.lexsort((dists, angles))
    sorted_pts = np.vstack([pivot, points[1:][order]])

    # Graham scan
    hull = []
    for p in sorted_pts:
        while len(hull) >= 2:
            # Cross product of (hull[-1]-hull[-2]) x (p-hull[-2])
            o = hull[-2]
            a = hull[-1]
            cross = (a[0]-o[0])*(p[1]-o[1]) - (a[1]-o[1])*(p[0]-o[0])
            if cross <= 0:  # clockwise or collinear → pop
                hull.pop()
            else:
                break
        hull.append(p)

    hull = np.array(hull)

    # Shoelace formula
    x, y = hull[:, 0], hull[:, 1]
    area = 0.5 * abs(np.dot(x, np.roll(y, -1)) - np.dot(y, np.roll(x, -1)))
    return area


def get_blob_edge(padded_binary_img: np.ndarray):

    edge1 = (padded_binary_img[:,1:] != padded_binary_img[:,:-1])
    edge1 = np.hstack([edge1, np.zeros((edge1.shape[0], 1), dtype=edge1.dtype)])

    edge2 = (padded_binary_img[1:,:] != padded_binary_img[:-1,:])
    edge2 = np.vstack([edge2, np.zeros((1, edge2.shape[1]), dtype=edge2.dtype)])
    
    return edge1 | edge2


def compute_fractal_dimension(binary_img: np.ndarray):

    # get only the blob's edge pixels
    img = get_blob_edge((binary_img > 0).astype(np.uint8))

    # box sizes as powers of 2
    min_dim = min(img.shape)
    scales  = [2**i for i in range(1, int(np.log2(min_dim)))]

    eps   = []
    boxes = []

    for s in scales:
        # tile the image with boxes of size s×s, count non-empty ones
        trimmed_h = (img.shape[0] // s) * s
        trimmed_w = (img.shape[1] // s) * s
        trimmed   = img[:trimmed_h, :trimmed_w]

        # reshape into s×s blocks and check if any pixel is set
        blocks = trimmed.reshape(trimmed_h // s, s, trimmed_w // s, s)
        count  = np.any(blocks, axis=(1, 3)).sum()

        eps.append(s)
        boxes.append(count)

    eps_axis  = np.log(eps)
    box_axis  = np.log(boxes)

    coeff = np.polyfit(eps_axis, box_axis, 1)
    return -coeff[0]  # slope is negative, so negate to get positive dimension


if __name__ == "__main__":

    import matplotlib.pyplot as plt

    # Initialize a 50x50 array of zeros
    blob_50x50 = np.zeros((50, 50), dtype=int)

    # Defined as (row, start_col, end_col)
    # Based on a 50x50 scale where the blob is centered/top-aligned
    rows_to_fill = [
    (6, 4, 34),                             # Top "overhang" row
    (7, 6, 34),
    (8, 5, 34),                             # Small indent
    (9, 6, 34),
    (10, 6, 34),
    (11, 6, 34),
    (12, 6, 34),
    (13, 6, 34),
    (14, 6, 34),
    (15, 6, 35),                            # Right-side bump
    (16, 6, 34),
    (17, 6, 34),
    (18, 6, 34),
    (19, 6, 34),
    (20, 6, 34),
    (21, 6, 34),
    (22, 6, 34),
    (23, 28, 34)                            # Bottom "leg" row
    ]

    # square
    # blob_50x50[15:35, 15:35] = 1

    # rectangle
    # blob_50x50[10:40, 15:35] = 1

    # line
    # blob_50x50[10:40, 15]    = 1

    for r, start, end in rows_to_fill:
        # end + 1 because slicing is exclusive
        blob_50x50[r, start:end+1] = 1

    plt.imshow(blob_50x50)
    plt.show()
    blob_array = np.pad(blob_50x50, 1, mode='constant', constant_values=0)

    hello = get_blob_edge(blob_array)
    plt.imshow(hello)
    plt.show()